File: Graph/BFS/14940.java
import java.util.*;
import java.io.*;


public class Main {
    private static int[] dx = {-1,1,0,0};
    private static int[] dy = {0,0,-1,1};
    private static int n,m;
    private static int[][] map,distance;
    private static boolean[][] visited;
    
    public static void main(String args[]) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    
        String[] size = br.readLine().split(" ");
        n = Integer.parseInt(size[0]);
        m = Integer.parseInt(size[1]);
        
        map = new int[n][m];
        distance = new int[n][m];
        visited = new boolean[n][m];

        int startX = -1,startY = -1;

        for(int i = 0; i < n; i++){
            map[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            if(startX == -1 && startY == -1){
                for(int j = 0; j < m; j++){
                    if(map[i][j] == 2){
                        startX = i;
                        startY = j;
                        break;
                    }
                }
            }
        }
        
        bfs(startX,startY);

        for(int i = 0; i < n; i++){
            for(int j =0; j < m; j++){
                if(!visited[i][j] && map[i][j] == 1){
                    bw.append(String.valueOf(-1) +" ");
                }else{
                    bw.append(String.valueOf(distance[i][j]) + " ");
                }
            }
            bw.append("\n");
        }
        
        bw.flush();
    }

    private static void bfs(int startX,int startY){
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(startX,startY));
        visited[startX][startY] = true;

        while(!q.isEmpty()){
            Pair cur = q.poll();

            for(int i = 0; i < 4; i++){
                int nextX = cur.a + dx[i];
                int nextY = cur.b + dy[i];

                if(nextX < 0 || nextY < 0 || nextX >= n || nextY >= m) continue;
                if(map[nextX][nextY] == 0) continue;
                if(visited[nextX][nextY]) continue;

                q.add(new Pair(nextX,nextY));
                distance[nextX][nextY] = distance[cur.a][cur.b] + 1;
                visited[nextX][nextY] = true;
            }
        }
    }
}

class Pair{
        public int a,b;
        public Pair(int a,int b){
            this.a = a;
            this.b = b;
        } 
    }

File: Graph/BFS/14940.cpp
#include <iostream>
#include <queue>

using namespace std;

int n,m;

int dx[] = {1,-1,0,0};
int dy[] = {0,0,1,-1};

int map[1001][1001],dist[1001][1001];
bool visited[1001][1001];
queue<pair<int,int>> q;

void bfs(){
    
    while(!q.empty()){
        int curX = q.front().first;
        int curY = q.front().second;
        q.pop();

        for(int i = 0; i < 4; i++){
            int nextX = curX + dx[i];
            int nextY = curY + dy[i];

            if(nextX < 0 || nextX >= n || nextY < 0 || nextY >= m) continue;
            if(map[nextX][nextY] == 0) continue;
            if(visited[nextX][nextY]) continue;

            q.push({nextX,nextY});
            dist[nextX][nextY] = dist[curX][curY] + 1;
            visited[nextX][nextY] = true;
        }

    }

}

int main(){
    cin >> n >> m;

    int startX = 0, startY = 0;

    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            cin >> map[i][j];
            
            if(map[i][j] == 2){
                startX = i;
                startY = j;
            }
        }
    }

    visited[startX][startY] = true;
    q.push({startX,startY});

    bfs();

    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            if(!visited[i][j] && map[i][j] == 1){
                cout << "-1 "; 
            }else{
                cout << dist[i][j] << " ";
            }
        }
        cout << "\n";
    }


    return 0;
}